yarn-project/p2p/src/testbench/testbench.ts

```diff
@@ -28,88 +27,305 @@ const testChainConfig: ChainConfig = {
+
+        readySignals.push(readySignal);
+        this.processes.push(childProcess);
+      }
+
+      // Wait for peers to all connect with each other
+      await sleep(4000);
+
+      // Wait for all peers to be booted up with timeout
+      await Promise.race([
+        Promise.all(readySignals),
+        new Promise((_, reject) =>
+          setTimeout(() => reject(new Error('Timeout waiting for all workers to be ready')), 30000),
+        ),
+      ]);
+
+      return this.peerEnrs;
+    } catch (error) {
+      // Clean up any processes that were created if there's an error
+      this.logger.error('Error during makeWorkerClients:', error);
+      await this.cleanup();
+      throw error;
+    }
+  }
+
+  purgeMessageReceivedByClient() {
+    this.messageReceivedByClient = new Array(this.processes.length).fill(0);
+  }
+
+  numberOfClientsThatReceivedMessage() {
+    return this.messageReceivedByClient.filter(count => count > 0).length;
+  }
+
+  /**
+   * Changes the port for a specific client
+   *
+   * @param clientIndex - The index of the client to change port for
+   * @param newPort - The new port to use
+   */
+  async changePort(clientIndex: number, newPort: number) {
+    try {
+      if (clientIndex < 0 || clientIndex >= this.processes.length) {
+        throw new Error(`Invalid client index: ${clientIndex}`);
+      }
+
+      this.processes[clientIndex].send({ type: 'STOP' });
+
+      // Wait for the process to be ready with a timeout
+      await sleep(1000);
+
+      this.logger.info(`Changing port for client ${clientIndex} to ${newPort}`);
+
+      // Update the port in the ports array
+      this.ports[clientIndex] = newPort;
+
+      // Update the port in the peerEnrs array
+      this.peerEnrs[clientIndex] = await makeEnr(this.peerIdPrivateKeys[clientIndex], newPort, testChainConfig);
+
+      // Maximum seed with 10 other peers to allow peer discovery to connect them at a smoother rate
+      const otherNodes = this.peerEnrs.filter(
+        (_, ind) => ind !== clientIndex && ind < Math.min(this.peerEnrs.length, 10),
+      );
+
+      const config = this.createClientConfig(clientIndex, newPort, otherNodes);
+      const [childProcess, readySignal] = this.spawnWorkerProcess(config, clientIndex);
+
+      this.processes[clientIndex] = childProcess;
+
+      // Wait for the process to be ready with a timeout
+      await Promise.race([
+        readySignal,
+        new Promise((_, reject) =>
+          setTimeout(() => reject(new Error(`Timeout waiting for client ${clientIndex} to be ready`)), 15000),
+        ),
+      ]);
+    } catch (error) {
+      this.logger.error(`Error during changePort for client ${clientIndex}:`, error);
+      // Only clean up the specific process that had an issue
+      await this.terminateProcess(this.processes[clientIndex], clientIndex);
+      throw error;
+    }
+  }
+
+  public getNewPort(): number {
+    while (true) {
```

Comment:
you can use the get-port package for this. It should already be in the package.json